### PR TITLE
Add support for stream block

### PR DESCRIPTION
**Pull request by jams2**
## Overview

This pull request adds a `StreamBlockFactory`, which supersedes the existing `StreamFieldFactory` functionality, adding supported for nested streams. Support for current functionality is maintained, and the new functionality uses the same syntax as `StreamFieldFactory` for instantiating a stream value's inner blocks (e.g. `MyPageFactory(body__0__struct__title="foobar")`.

With the new functionality, `StreamBlockFactory` subclasses can be defined and used as follows:

```python
class StreamBlockFactoryA(StreamBlockFactory):
    struct = factory.SubFactory(MyStructBlock)
    text = factory.SubFactory(MyTextBlock)

    class Meta:
        model = StreamBlockA


class MyPageFactory(PageFactory):
    body = StreamFieldFactory(StreamBlockFactoryA)

    class Meta:
        model = MyPage
```

`StreamBlockFactories` can be used as sub-factories of `StructBlockFactory`, `ListBlockFactory` and `StreamBlockFactory`, for example:

```python
class StreamBlockFactoryB(StreamBlockFactory):
    inner_stream = factory.SubFactory(StreamBlockFactoryA)

    class Meta:
        model = StreamBlockB
```

## Breaking changes

This PR significantly changes the way that ``StreamFieldFactory`` works under the hood. The changes provide a deeper integration with factory_boy's object generation logic, resulting in the requirement that all `BlockFactory` values passed to an old-style `StreamFieldFactory` (dictionary-style) declaration must be wrapped in `SubFactories` (this is covered in more detail in the changes to the readme). I propose a new major version release (3.0)

### CODE REVIEW

File: src/wagtail_factories/blocks.py [hunk 1]
```diff
@@ -56,15 +51,64 @@ def generate(self, step, params):
...
+        stream_data = []
+        for index, block_items in sorted(indexed_mappings.items()):
             for block_name, block_params in block_items.items():
-                block_factory = self.get_factory_for_block(block_name)
-                value = block_factory(**block_params)
-                retval.append((block_name, value))
-        return retval
+                block_factory = cls._get_block_factory(block_name)
+                if isinstance(block_factory, ListBlockFactory):
+                    stream_data.append((block_name, block_factory(**block_params)))
```
Comment (jams2) [inline src/wagtail_factories/blocks.py:65]: @bcdickinson Tests are passing but I think we need to pass the strategy through to ListBlockFactories as well, so they can propagate that to their children when we're explicitly instantiating the ListBlockFactory. Any thoughts on how to approach this?

File: src/wagtail_factories/blocks.py [hunk 2]
```diff
@@ -142,17 +186,56 @@ class StructBlockFactory(factory.Factory):
...
-                (name, (kwargs[name] if name in kwargs else child_block.get_default()))
-                for name, child_block in block.child_blocks.items()
-            ],
-        )
+        return cls._generate(factory.enums.BUILD_STRATEGY, kwargs)
 
     @classmethod
     def _create(cls, model_class, *args, **kwargs):
-        return cls._build(model_class, *args, **kwargs)
+        return cls._generate(factory.enums.CREATE_STRATEGY)
```
Comment (bcdickinson) [inline src/wagtail_factories/blocks.py:241]: Should this pass `kwargs` as well?

**Review by bcdickinson — COMMENTED** [review]
This is looking really promising, but in testing this on a project I've been working on, I found that having `LazyFunction` declarations on `StructBlockFactory` subclasses is broken by this, so further work is needed.

File: src/wagtail_factories/blocks.py [hunk 3]
```diff
@@ -56,15 +51,64 @@ def generate(self, step, params):
                     continue
 
                 index = int(index)
-                result[index][block_name][param] = value
+                mappings[index][block_name][param] = value
+        return mappings
```
[the remark this answers is not in the document]
Comment (jams2) [reply src/wagtail_factories/blocks.py:55]: The error that is most likely being caught here is not getting enough values to unpack, so a key with no dunders, which would be incorrect: the arguments passed to the `StreamBlockFactory` initialisation call must have at least an index and a block name (e.g. `0__foo_block`). I think we should raise an error in that case.

[the remark this answers is not in the document]
Comment (jams2) [reply src/wagtail_factories/blocks.py:55]: There's another case that's not accounted for. A `StreamBlock`'s child block might be an atomic block (e.g. an `IntegerBlock`), which could be declared like `0__a_number=42`, which would also cause a `ValueError` in this case.

File: src/wagtail_factories/blocks.py [hunk 4]
```diff
@@ -142,17 +186,56 @@ class StructBlockFactory(factory.Factory):
...
+            if isinstance(declaration, ListBlockFactory):
+                return lambda **params: declaration(**params)
+            elif isinstance(declaration, factory.Factory):
+                return lambda **params: declaration.generate(strategy, **params)
+            elif isinstance(declaration, factory.SubFactory):
+                wrapped_factory = declaration.get_factory()
+                return lambda **params: wrapped_factory.generate(strategy, **params)
+            else:
+                # It's either a scalar or a non-factory callable
+                return lambda: declaration
```
[the remark this answers is not in the document]
Comment (jams2) [reply src/wagtail_factories/blocks.py]: Fixed in the latest batch of work

**Comment by jams2** [comment]
Thanks for your initial review @bcdickinson ! I've marked the items that are no longer relevant as resolved.

File: src/wagtail_factories/blocks.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bcdickinson) [reply src/wagtail_factories/blocks.py]: This is fixed in my local testing and the tests look good too. Nice work!

**Comment by bcdickinson** [comment]
I'm encountering an issue testing this locally when calling factory classes that contain `StreamFieldFactory` declarations and not passing any kwargs (e.g. just `MyPageFactory(parent=some_page)` with a `body__0="my_struct_block"` kwarg or some such). It seems like the step builder code can't handle an absence of block data and I get this (truncated stack trace to only show factory code frames):

```
  File "/venv/lib/python3.8/site-packages/factory/base.py", line 40, in __call__
    return cls.create(**kwargs)
  File "/venv/lib/python3.8/site-packages/factory/base.py", line 528, in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
  File "/venv/lib/python3.8/site-packages/factory/django.py", line 117, in _generate
    return super()._generate(strategy, params)
  File "/venv/lib/python3.8/site-packages/factory/base.py", line 465, in _generate
    return step.build()
  File "/venv/lib/python3.8/site-packages/factory/builder.py", line 258, in build
    step.resolve(pre)
  File "/venv/lib/python3.8/site-packages/factory/builder.py", line 199, in resolve
    self.attributes[field_name] = getattr(self.stub, field_name)
  File "/venv/lib/python3.8/site-packages/factory/builder.py", line 344, in __getattr__
    value = value.evaluate_pre(
  File "/venv/lib/python3.8/site-packages/factory/declarations.py", line 48, in evaluate_pre
    return self.evaluate(instance, step, context)
  File "/venv/lib/python3.8/site-packages/wagtail_factories/blocks.py", line 113, in evaluate
    return self.stream_block_factory(**extra)
  File "/venv/lib/python3.8/site-packages/factory/base.py", line 40, in __call__
    return cls.create(**kwargs)
  File "/venv/lib/python3.8/site-packages/factory/base.py", line 528, in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
  File "/venv/lib/python3.8/site-packages/wagtail_factories/blocks.py", line 44, in _generate
    step = cls._builder_class(cls._meta, params, strategy)
  File "/venv/lib/python3.8/site-packages/wagtail_factories/builder.py", line 46, in __init__
    indexed_block_names, extra_declarations = self.get_block_declarations(
  File "/venv/lib/python3.8/site-packages/wagtail_factories/builder.py", line 100, in get_block_declarations
    self.validate_block_indexes(indexed_block_names, factory_meta)
  File "/venv/lib/python3.8/site-packages/wagtail_factories/builder.py", line 108, in validate_block_indexes
    for declared, expected in zip_longest(indexes, range(max(indexes) + 1)):
ValueError: max() arg is an empty sequence```

As discussed on our call just now, I think in this instance it would be sensible to just handle this by leaving the body empty and not throwing an error. Also, as we noted, there's currently no way of defining default data for the `StreamBlockFactory`, so I'll raise a new issue for that.

**Comment by jams2** [comment]
@bcdickinson good catch, the same issue also existed with requesting the default value for a StreamBlockFactory in a StreamBlockFactory, e.g.

```python
MyPageFactory(body__0="inner_stream")
```

Fixed in https://github.com/wagtail/wagtail-factories/pull/55/commits/e893a8a47f421437239c288ce1d493b87000d8d8

**Comment by bcdickinson** [comment]
Nice, I'd say this is there now! Happy for this to be merged and released.

**Pull request merged**